sdk/appcenter-distribute/src/main/java/com/microsoft/appcenter/distribute/download/http/HttpConnectionReleaseDownloader.java

```diff
@@ -0,0 +1,244 @@
+        if (downloadedReleaseFilePath != null) {
+
+            /* Check if it's the same release. */
+            File downloadedReleaseFile = new File(downloadedReleaseFilePath);
+            if (downloadedReleaseFilePath.equals(targetFile.getAbsolutePath())) {
+                if (downloadedReleaseFile.exists()) {
+                    onDownloadComplete(targetFile);
+                    return;
+                }
+            } else {
+
+                /* Remove previously downloaded release. */
+                removeFile(downloadedReleaseFile);
+                SharedPreferencesManager.remove(PREFERENCE_KEY_DOWNLOADED_RELEASE_FILE);
+            }
+        }
+        if (!NetworkStateHelper.getSharedInstance(mContext).isNetworkConnected()) {
+            mListener.onError("No network connection, abort downloading.");
+            return;
+        }
+        String[] permissions = requiredPermissions();
+        int[] permissionsState = PermissionUtils.permissionsState(mContext, permissions);
+        if (!PermissionUtils.permissionsAreGranted(permissionsState)) {
+            mListener.onError("No external storage permission.");
+            return;
+        }
+        if (!isDownloading()) {
+
+            /* Start download the release package file. */
+            long enqueueTime = System.currentTimeMillis();
+            downloadFile(targetFile);
+            mListener.onStart(enqueueTime);
+            showProgressNotification(0, 0);
+        } else {
+            AppCenterLog.debug(LOG_TAG, "Downloading of " + targetFile.getPath() + " is already in progress.");
+        }
+    }
+
+    @Override
+    public synchronized void cancel() {
+        if (mHttpDownloadFileTask != null) {
+            mHttpDownloadFileTask.cancel(true);
+            mHttpDownloadFileTask = null;
+        }
+        String filePath = SharedPreferencesManager.getString(PREFERENCE_KEY_DOWNLOADED_RELEASE_FILE, null);
+        if (filePath != null) {
+            removeFile(new File(filePath));
+            SharedPreferencesManager.remove(PREFERENCE_KEY_DOWNLOADED_RELEASE_FILE);
+        }
+        cancelProgressNotification();
+    }
+
+    private void downloadFile(File file) {
+        Uri downloadUrl = mReleaseDetails.getDownloadUrl();
+        AppCenterLog.debug(LOG_TAG, "Start downloading new release from " + downloadUrl);
+        mHttpDownloadFileTask = AsyncTaskUtils.execute(LOG_TAG, new HttpDownloadFileTask(this, downloadUrl, file));
+    }
+
+    private void removeFile(File file) {
+        AppCenterLog.debug(LOG_TAG, "Removing downloaded file from " + file.getAbsolutePath());
+        AsyncTaskUtils.execute(LOG_TAG, new RemoveFileTask(file));
+    }
+
+    private void showProgressNotification(long currentSize, long totalSize) {
+        Notification.Builder builder = getNotificationBuilder();
+
+        /* TODO: We should not reuse "mandatory" string here. */
```

Comment:
Alias the resource without the mandatory in the key as the text suits optional too currently.
Comment about deprecation of the resource in changelog.